python/dolma/core/parallel.py

```diff
@@ -199,36 +301,61 @@ def process_single(
         raise NotImplementedError()
 
     @classmethod
-    def _process_single_and_save_status(
+    def _log_backoff(cls, details: Details):
+        """Log backoff details."""
+        message = (
+            f"Backing off `{details['target'].__name__}` "
+            f"after {details['tries']:,} "
+            f"tries (wait: {details.get('wait', 0.0):.2f}s)"
+        )
+        if ex := details.get("exception"):
+            # add details about the exception to the message
+            import traceback  # pylint: disable=import-outside-toplevel
+
+            message += " due to " + "".join(traceback.format_exception_only(type(ex), ex)).strip()
+
+        cls.get_logger().warning(message)
+
+    @classmethod
+    def _process_batch_and_save_status(
         cls,
-        source_path: str,
-        destination_path: str,
-        metadata_path: str,
+        source_paths: List[str],
+        destination_paths: List[str],
+        metadata_paths: List[str],
         queue: QueueType,
-        serialized_kwargs: bytes,
+        serialized_kwargs: List[bytes],
+        backoff_max_time: float,
+        backoff_max_tries: int,
+        backoff_exceptions: Tuple[Type[Exception], ...],
     ):
         """A wrapper around process single that saves a metadata file if processing is successful."""
 
         # make destination directory if it doesn't exist for the destination and metadata paths
-        mkdir_p(parent(destination_path))
-        mkdir_p(parent(metadata_path))
-
-        kwargs = pickle.loads(serialized_kwargs)
-        retries_on_error = kwargs.get("retries_on_error", 0) + 1
-        while True:
-            try:
-                cls.process_single(
-                    source_path=source_path, destination_path=destination_path, queue=queue, **kwargs
-                )
-                break
-            except DolmaRetryableFailure as exception:
-                retries_on_error -= 1
-                if retries_on_error == 0:
-                    raise DolmaError from exception
+        for path in itertools.chain(destination_paths, metadata_paths):
+            mkdir_p(parent(path))
+
+        # we unpickle the serialized kwargs
+        deserialized_kwargs = [pickle.loads(kw) for kw in serialized_kwargs]
```

Comment:
Not sure if an issue for you but consider that this use of pickle.loads() might be a memory hog.